lib/fizzy/execute.cpp

```diff
@@ -555,13 +556,33 @@ inline constexpr T fmax(T a, T b) noexcept
     return a < b ? b : a;
 }
 
-__attribute__((no_sanitize("float-cast-overflow"))) inline constexpr float demote(
-    double value) noexcept
+__attribute__((no_sanitize("float-cast-overflow"))) inline float demote(double value) noexcept
 {
+    // Save current rounding mode.
+    // Negative value denote error. On architectures without rounding mode support glibc always
+    // returns FE_TONEAREST, but there are C standard library implementations that return
+    // error code, e.g. -ENOTSUP.
+    const auto current_rounding_mode = std::fegetround();
+
+    // Fast path if the rounding more is the desired or could not be determined.
+    if (current_rounding_mode == FE_TONEAREST)
+        return static_cast<float>(value);
+
+    // Change rounding mode to "round to nearest tie-to-even" (the default IEEE 754 mode).
+    // The returned status code is ignored because it means the input value is invalid (unlikely)
+    // or the target architecture does not support changing rounding mode.
+    std::fesetround(FE_TONEAREST);
+
     // The float-cast-overflow UBSan check disabled for this conversion. In older clang versions
     // (up to 8.0) it reports a failure when non-infinity f64 value is converted to f32 infinity.
     // Such behavior is expected.
-    return static_cast<float>(value);
+    // The result is volatile to enforce it being computed before the ending std::fesetround().
+    const volatile auto result = static_cast<float>(value);
```

Comment:
Does this `volatile` trick always guarantee the right order?

Reply:
Hard to tell.

Reply:
Couldn't you instead add a branch to trigger different layout?

```
if (current_rounding_mode  != status_code || status_code == -1) std::fesetround(current_rounding_mode);
```

Reply:
Getting to a point inline assembly would be easier/nicer here for the entire function lol.

Reply:
> Couldn't you instead add a branch to trigger different layout?
> 
> ```
> if (current_rounding_mode  != status_code || status_code == -1) std::fesetround(current_rounding_mode);
> ```

This is done some comments earlier.